File: apps/api/pkg/models/billing.go
package models

import (
	"context"
	"encoding/json"
	"time"

	"gorm.io/datatypes"
)

// Bill & Payment (M20 — DEV_SPEC §4.2).
//
// One table holds every purchase, whatever was sold. `kind` says what it was,
// `period_start`/`period_end` are filled in only for a subscription invoice,
// and `provider`/`provider_ref` are where a real gateway will write its charge
// id. Adding monthly billing therefore adds rows, not columns.

// Order is a purchase and its receipt — the same record, because they are the
// same fact. Once written it is never edited: a refund is a status change and a
// correction is a new order, so that a receipt a user has already downloaded
// cannot quietly become a different document.
type Order struct {
	Base
	UserID string `gorm:"type:char(36);not null;index:idx_order_user_issued,priority:1" json:"user_id"`
	// Human-facing receipt number, "RV-2569-000123". Unique so a race that
	// produced the same sequence twice fails loudly instead of printing two
	// receipts with one number.
	Number string `gorm:"type:varchar(24);not null;uniqueIndex" json:"number"`
	Kind   string `gorm:"type:varchar(24);not null;default:'ai_credit'" json:"kind"`
	Status string `gorm:"type:varchar(12);not null;default:'paid'" json:"status"`
	Title  string `gorm:"type:varchar(160);not null" json:"title"`
	// The lines, as JSON rather than a child table. A receipt is an immutable
	// document that is only ever read whole; a join buys nothing and a schema
	// change to `order_lines` would rewrite receipts that were already issued.
	Lines datatypes.JSON `gorm:"type:json" json:"lines"`

	SubtotalTHB float64 `gorm:"type:decimal(12,2);not null;default:0" json:"subtotal_thb"`
	DiscountTHB float64 `gorm:"type:decimal(12,2);not null;default:0" json:"discount_thb"`
	TotalTHB    float64 `gorm:"type:decimal(12,2);not null;default:0" json:"total_thb"`
	Currency    string  `gorm:"type:char(3);not null;default:'THB'" json:"currency"`

	Method      string `gorm:"type:varchar(16);not null;default:'card'" json:"method"`
	MethodLabel string `gorm:"type:varchar(60)" json:"method_label"`
	// Points debited for this order. Points are not baht and are never added to
	// one — they are reported side by side.
	PointsSpent int `gorm:"not null;default:0" json:"points_spent"`

	// The trip the purchase was used on, when it was bought for one. Nullable
	// and never a foreign key: deleting a trip must not delete its receipts.
	TripID    *string `gorm:"type:char(36);index" json:"trip_id"`
	TripTitle string  `gorm:"type:varchar(160)" json:"trip_title"`

	Provider    string `gorm:"type:varchar(24)" json:"provider"`
	ProviderRef string `gorm:"type:varchar(120)" json:"provider_ref"`
	// True while there is no gateway behind a cash order (§16).
	Simulated bool `gorm:"not null;default:false" json:"simulated"`

	SubscriptionID *string    `gorm:"type:char(36);index" json:"subscription_id"`
	PeriodStart    *time.Time `json:"period_start"`
	PeriodEnd      *time.Time `json:"period_end"`

	IssuedAt   time.Time  `gorm:"not null;index:idx_order_user_issued,priority:2" json:"issued_at"`
	PaidAt     *time.Time `json:"paid_at"`
	RefundedAt *time.Time `json:"refunded_at"`
}

func (Order) TableName() string { return "orders" }

// OrderLine is one row of a receipt. Stored inside Order.Lines as JSON.
type OrderLine struct {
	Label         string  `json:"label"`
	Quantity      int     `json:"quantity"`
	UnitAmountTHB float64 `json:"unit_amount_thb"`
	AmountTHB     float64 `json:"amount_thb"`
}

// DecodeOrderLines reads the JSON column back. A receipt that cannot be parsed
// renders as no lines rather than as an error page: the totals beside it are
// stored as columns and are still true.
func DecodeOrderLines(raw datatypes.JSON) []OrderLine {
	if len(raw) == 0 {
		return nil
	}
	var out []OrderLine
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil
	}
	return out
}

// EncodeOrderLines is the other direction, for the handler that writes one.
func EncodeOrderLines(lines []OrderLine) datatypes.JSON {
	raw, err := json.Marshal(lines)
	if err != nil {
		return datatypes.JSON("[]")
	}
	return datatypes.JSON(raw)
}

// Subscription statuses.
const (
	SubStatusActive   = "active"
	SubStatusPastDue  = "past_due"
	SubStatusCanceled = "canceled"
)

// Subscription is a standing plan. No row exists for a free user — the API
// synthesises the free plan rather than writing a row per signup, so this table
// only ever holds people who are actually being charged.
type Subscription struct {
	Base
	UserID   string  `gorm:"type:char(36);not null;index" json:"user_id"`
	PlanID   string  `gorm:"type:varchar(40);not null" json:"plan_id"`
	Status   string  `gorm:"type:varchar(12);not null;default:'active'" json:"status"`
	Interval string  `gorm:"type:varchar(8);not null;default:'month'" json:"interval"`
	PriceTHB float64 `gorm:"type:decimal(12,2);not null;default:0" json:"price_thb"`

	CurrentPeriodStart time.Time `gorm:"not null" json:"current_period_start"`
	CurrentPeriodEnd   time.Time `gorm:"not null" json:"current_period_end"`
	// Set when the user has cancelled but the paid period is still running.
	CancelAtPeriodEnd bool `gorm:"not null;default:false" json:"cancel_at_period_end"`

	Provider    string `gorm:"type:varchar(24)" json:"provider"`
	ProviderRef string `gorm:"type:varchar(120)" json:"provider_ref"`

	CanceledAt *time.Time `json:"canceled_at"`
}

func (Subscription) TableName() string { return "subscriptions" }

// BillingSummary is computed, never stored: a purchase total that is kept in a
// column is a purchase total that can disagree with the receipts.
type BillingSummary struct {
	Orders            int
	AIDraftsPurchased int
	TotalSpentTHB     float64
	PointsSpent       int
	Since             *time.Time
}

type BillingStore interface {
	// CreateOrder assigns the receipt number and writes the order in one
	// transaction, so two concurrent purchases cannot share a number.
	CreateOrder(ctx context.Context, order *Order) error
	ListOrders(ctx context.Context, userID string, limit int) ([]Order, error)
	GetOrder(ctx context.Context, userID, orderID string) (*Order, error)
	Summary(ctx context.Context, userID string) (*BillingSummary, error)

	// ActiveSubscription returns nil (with no error) for a free user.
	ActiveSubscription(ctx context.Context, userID string) (*Subscription, error)
	SaveSubscription(ctx context.Context, sub *Subscription) error
}

File: apps/web/lib/data/mock/billing.ts
import { FREE_DRAFTS_PER_TRIP, FREE_PLAN_ID } from '@/lib/catalog/plans';

import { AI_CREDITS } from './seed/trip';

import type { Order, OrderKind, PayChannel, PaymentMethod, Subscription } from '../types';

/**
 * Bill & Payment (M20) — the mock side.
 *
 * Two things live here: the price list, which is reference data the API owns in
 * live mode, and the receipt bookkeeping the mock repository needs in order to
 * behave like a real one (numbering, totals, the seeded history a UAT tester
 * opens the screen expecting to see).
 *
 * The plan catalogue itself moved to `lib/catalog/plans.ts` in M26 and is
 * re-exported below, so mock mode and the public pricing page read one list.
 */

/**
 * What a paid draft can be paid with. Listed in the paywall so nobody has to
 * tap "ซื้อ" to find out whether their method is accepted — and the id, not the
 * label, is what the receipt records.
 */
export const AI_PAY_CHANNELS: PayChannel[] = [
  { id: 'card', label: 'บัตรเครดิต/เดบิต' },
  { id: 'promptpay', label: 'พร้อมเพย์ (QR)' },
  { id: 'truemoney', label: 'TrueMoney Wallet' },
];

/**
 * The catalogue lives in `lib/catalog/plans.ts` — one copy on this side of the
 * wire, shared with the public pricing page, which cannot fetch it because the
 * catalogue endpoint is behind sign-in.
 */
export {
  FREE_ACTIVE_TRIPS,
  FREE_PLAN_ID,
  PLANS,
  ROVE_YEAR_PLAN_ID,
  TRIP_PASS_PLAN_ID,
  UNLIMITED_DRAFTS,
} from '@/lib/catalog/plans';

/**
 * Everyone is on the free plan until a gateway exists. It is a real
 * subscription record in shape — status, period, interval — so the screen that
 * renders a paid one is already the screen being rendered today.
 */
export const FREE_SUBSCRIPTION: Subscription = {
  id: null,
  planId: FREE_PLAN_ID,
  planName: 'ROVE ฟรี',
  status: 'none',
  interval: null,
  priceThb: 0,
  currentPeriodStart: null,
  currentPeriodEnd: null,
  cancelAtPeriodEnd: false,
  includedDraftsPerPeriod: FREE_DRAFTS_PER_TRIP,
};

/* ------------------------------------------------------------- numbering -- */

/**
 * "RV-2569-000004" — Buddhist year, then a per-year sequence.
 *
 * A receipt number is what a person quotes when they write in about a charge,
 * so it is short, spoken out loud without ambiguity, and never reused. Mock
 * mode counts the orders it already has for that year; the API does the same
 * count inside a transaction and leans on a unique index (§4.2).
 */
export function receiptNumber(issuedAt: string, existing: Order[]) {
  const year = new Date(issuedAt).getFullYear() + 543;
  const prefix = `RV-${year}-`;
  const taken = existing.filter((o) => o.number.startsWith(prefix)).length;
  return `${prefix}${String(taken + 1).padStart(6, '0')}`;
}

/* ---------------------------------------------------------------- orders -- */

export interface DraftOrderInput {
  kind: OrderKind;
  title: string;
  quantity: number;
  unitAmountThb: number;
  method: PaymentMethod;
  methodLabel: string;
  lineLabel: string;
  pointsSpent?: number;
  tripId?: string | null;
  tripTitle?: string | null;
  issuedAt: string;
}

/**
 * Builds a completed order. Points and cash produce the *same* record — one
 * with a total of ฿0 and a points figure, one the other way round — because
 * "what have I spent on ROVE" is a question about both.
 */
export function buildOrder(id: string, input: DraftOrderInput, existing: Order[]): Order {
  const amount = input.unitAmountThb * input.quantity;
  const paidWithPoints = input.method === 'points';

  return {
    id,
    number: receiptNumber(input.issuedAt, existing),
    kind: input.kind,
    status: 'paid',
    title: input.title,
    lines: [
      {
        label: input.lineLabel,
        quantity: input.quantity,
        unitAmountThb: input.unitAmountThb,
        amountThb: amount,
      },
    ],
    subtotalThb: amount,
    // Points do not reduce a price, they replace it: the line keeps its list
    // price and the whole of it is discounted away, so the receipt still says
    // what the thing was worth.
    discountThb: paidWithPoints ? amount : 0,
    totalThb: paidWithPoints ? 0 : amount,
    currency: 'THB',
    method: input.method,
    methodLabel: input.methodLabel,
    pointsSpent: input.pointsSpent ?? 0,
    tripId: input.tripId ?? null,
    tripTitle: input.tripTitle ?? null,
    provider: null,
    providerRef: null,
    // No gateway in Phase 1: a cash order is recorded, not charged.
    simulated: !paidWithPoints,
    periodStart: null,
    periodEnd: null,
    issuedAt: input.issuedAt,
    paidAt: input.issuedAt,
    refundedAt: null,
  };
}

/** What the old per-draft product cost, before M26 withdrew it. */
const LEGACY_DRAFT_PRICE_THB = 39;

/**
 * The history a UAT tester finds already there: one draft bought back when
 * drafts were sold one at a time, and one Trip Pass that came back because the
 * trip it unlocked ended in a booking.
 *
 * The withdrawn product is in the list on purpose. Receipts are a record of
 * what happened, not of what is currently for sale, and a billing screen that
 * cannot render its own history is one that would have to be repaired every
 * time the price list changes.
 */
export function seedOrders(): Order[] {
  const out: Order[] = [];

  out.push(
    buildOrder(
      'ord_seed_1',
      {
        kind: 'ai_credit',
        title: 'ร่างแพลนด้วย AI เพิ่ม 1 ครั้ง',
        lineLabel: 'สิทธิ์ให้ AI ร่างแพลน (ทริปญี่ปุ่นใบไม้เปลี่ยนสี 2569)',
        quantity: 1,
        unitAmountThb: LEGACY_DRAFT_PRICE_THB,
        method: 'promptpay',
        methodLabel: 'พร้อมเพย์ (QR)',
        tripId: 'demo',
        tripTitle: 'ญี่ปุ่นใบไม้เปลี่ยนสี 2569',
        issuedAt: '2026-05-12T09:20:00.000Z',
      },
      out,
    ),
  );

  const refunded = buildOrder(
    'ord_seed_2',
    {
      kind: 'trip_pass',
      title: 'Trip Pass — โซลกับเพื่อนสนิท',
      lineLabel: 'ปลดล็อกทริป (ให้ AI ร่างและปรับแพลนไม่จำกัด)',
      quantity: 1,
      unitAmountThb: AI_CREDITS.passPriceThb,
      method: 'promptpay',
      methodLabel: 'พร้อมเพย์ (QR)',
      tripId: 'seoul',
      tripTitle: 'โซลกับเพื่อนสนิท',
      issuedAt: '2026-01-20T02:05:00.000Z',
    },
    out,
  );
  // Booked through ROVE on 26 Jan, so the pass was paid back (A26.4). The
  // status changes and the row stays: a refund is a new fact about an order,
  // never a receipt quietly rewritten.
  refunded.status = 'refunded';
  refunded.refundedAt = '2026-01-26T09:25:00.000Z';
  out.push(refunded);

  return out;
}
